File: articles/python-chemistry-simulation-laboratory-data/julia/python_chemistry_kernel.jl
# Python chemistry numerical kernel in Julia.
# This mirrors simple calculations used in Python-centered chemistry workflows.

function unknown_concentration(response, slope, intercept)
    return (response - intercept) / slope
end

function first_order_concentration(c0, k, t)
    return c0 * exp(-k * t)
end

function half_life_first_order(k)
    return log(2.0) / k
end

function standard_error(sd, n)
    return sd / sqrt(n)
end

println("unknown_concentration_mM=", round(unknown_concentration(0.95, 0.30, 0.02), digits=6))
println("first_order_concentration_mM=", round(first_order_concentration(10.0, 0.015, 100.0), digits=6))
println("half_life_s=", round(half_life_first_order(0.015), digits=6))
println("standard_error=", round(standard_error(0.03, 3), digits=6))

File: articles/python-chemistry-simulation-laboratory-data/fortran/python_chemistry_kernel.f90
program python_chemistry_kernel
  implicit none

  real(8) :: response, slope, intercept, unknown
  real(8) :: c0, k, t, concentration, half_life
  real(8) :: sd, n, se

  response = 0.95d0
  slope = 0.30d0
  intercept = 0.02d0
  unknown = (response - intercept) / slope

  c0 = 10.0d0
  k = 0.015d0
  t = 100.0d0
  concentration = c0 * exp(-k * t)
  half_life = log(2.0d0) / k

  sd = 0.03d0
  n = 3.0d0
  se = sd / sqrt(n)

  print *, "unknown_concentration_mM=", unknown
  print *, "first_order_concentration_mM=", concentration
  print *, "half_life_s=", half_life
  print *, "standard_error=", se

end program python_chemistry_kernel
